docs/tutorial/write-a-machine-charm-for-tinyproxy.md

```diff
@@ -0,0 +1,577 @@
+
+TODO: Multipass
+
+### Install Juju and charm development tools
+
+TODO: Charmcraft, Juju, LXD (all using Concierge)
+
+### Install Python development tools
+
+TODO: tox, uv
+
+## Create a charm project
+
+You'll need to work inside the virtual machine again. Open a terminal, then run:
+
+```text
+multipass shell juju-sandbox
+```
+
+The terminal prompt is now `ubuntu@juju-sandbox:~$`.
+
+Next, create a project directory and use Charmcraft to create the initial version of your charm:
+
+```text
+mkdir tinyproxy
+cd tinyproxy
+charmcraft init --profile machine
+```
+
+TODO:
+
+- Briefly summarise the most important files that Charmcraft creates.
+- Edit charmcraft.yaml to set the title, description, summary, and platform.
+- Suggest using nano to edit files.
+
+## Write your charm
+
+### Write a helper module
+
+Your charm will interact with tinyproxy, so it's a good idea to start by writing a helper module that wraps tinyproxy.
+Charmcraft created `src/tinyproxy.py` as a placeholder helper module.
+
+The helper module will be independent of the main logic of your charm. This will make it easier to test your charm. However, the helper module won't be a general-purpose wrapper for tinyproxy. The helper module will contain opinionated functions for managing tinyproxy on Ubuntu.
+
+The helper module will depend on some libraries that are useful when writing charms. Add the libraries to your charm's dependencies:
+
+```text
+uv add charmlibs-apt charmlibs-pathops
+```
+
+This has added the following Python packages to the `dependencies` list in `pyproject.toml`:
+
+- `charmlibs-apt` - A library for using APT to manage system packages. This is how your charm will install tinyproxy.
+- `charmlibs-pathops` - A file operations library, similar to `pathlib` from the standard library.
+
+Next, replace the contents of `src/tinyproxy.py` by:
+
+```python
+import logging
+import os
+import re
+import shutil
+import signal
+import subprocess
+
+from charmlibs import apt, pathops
+
+logger = logging.getLogger(__name__)
+
+CONFIG_FILE = "/etc/tinyproxy/tinyproxy.conf"
+PID_FILE = "/var/run/tinyproxy.pid"
+
+
+def check_slug(slug: str) -> None:
+    """Check that the URL slug is valid. Raise ValueError otherwise."""
+    if not re.fullmatch(r"[a-z0-9-]+", slug):
+        raise ValueError(f"Invalid slug: '{slug}'. Slug must match the regex [a-z0-9-]+")
+
+
+def ensure_config(port: int, slug: str) -> bool:
+    """Ensure that tinyproxy is configured. Return True if any changes were made."""
+    # For the config file format, see https://manpages.ubuntu.com/manpages/jammy/en/man5/tinyproxy.conf.5.html
+    config = f"""\
+PidFile "{PID_FILE}"
+Port {port}
+Timeout 600
+ReverseOnly Yes
+ReversePath "/{slug}/" "http://www.example.com/"
+"""
+    return pathops.ensure_contents(CONFIG_FILE, config)
+
+
+def install() -> None:
+    """Use APT to install the tinyproxy executable."""
+    apt.update()
+    try:
+        apt.add_package("tinyproxy-bin")
```

Comment:
I wonder if this should install a specific version? For the tutorial it's probably more convenient to not do that, and not needed. However, it does seem to be what most machine charms do, to make their upgrade and QA processes simpler.

(But not the config for installing different versions like the existing tutorial - that seems a strange thing to do).